### PR TITLE
Opt out

### LINKED ISSUE #537: Visitor app: opt out handling
admin needs a way to mark visitors as opted out when they originally opted into marketing and sms

### PR DESCRIPTION
**Pull request by eliascodes**
Fixes #537 

Related to #680 

#### Changes
* Create a new simpler/better custom checkbox
* Use it on the visitor details page to allow contact consent to be changed

Manual testing:
* Change contact consent of a visitor and press save

### CODE REVIEW

**Comment by eliascodes** [comment]
Note:
I've not been able to think of or implement a satisfactory layout for the checkboxes, so am welcoming reccommendations.

Stacked:
screenshot 2019-02-26 at 22 11 39

Inline, desktop:
screenshot 2019-02-27 at 08 41 18

Inline, iPad:
screenshot 2019-02-27 at 08 41 07

**Comment by astroash** [comment]
@eliascodes stacked in my preference. I think it needs either centering or a bit more top margin.

**Comment by eliascodes** [comment]
Both these examples have a bit more top margin.

Horizontal centering is imperfect because of text content. I'm eyeballing it at the moment:

Centered(ish):
screenshot 2019-02-27 at 13 51 55

No horizontal centering
screenshot 2019-02-27 at 13 50 20

Of the two I prefer left-aligned.

**Comment by astroash** [comment]
Can you post a full screen pic of the left align with top margin?

**Comment by eliascodes** [comment]
screenshot 2019-02-28 at 21 23 41

**Comment by astroash** [comment]
@eliascodes is it possible for the we buttons to also line up with the width of the input boxes? That’s the only thing that is throwing it off for me atm. Margin looks 👌🏽

**Comment by astroash** [comment]
Also is this ready for review/merge or was assignment for the screenshot?

**Comment by eliascodes** [comment]
> is it possible for the we buttons to also line up with the **width** of the input boxes

Assuming you mean _height_ here

> Also is this ready for review/merge or was assignment for the screenshot?

Both.

**Comment by eliascodes** [comment]
You mean like this?

screenshot 2019-03-01 at 09 56 35

File: src/cb_admin/components/QrBox.js [hunk 1]
```diff
@@ -37,18 +35,21 @@ const QrBox = ({ qrCodeUrl, print, send, error, status }) => {
...
-      <Button onClick={print}>PRINT QR CODE</Button>
-      <Button onClick={send} disabled={hasSent}>
-        {
-          hasSent
-            ? 'QR CODE SENT'
-            : 'RESEND QR CODE'
-        }
-      </Button>
+      <Row between={'xs'}>
+        <Button onClick={print}>PRINT QR CODE</Button>
```
Comment (astroash) [inline src/cb_admin/components/QrBox.js]: Can we remove line 18:

```js
 margin: 0 2.5%;
```

This is what is making the buttons not quite in line with the rest of the page.

File: src/cb_admin/pages/Visitor.js [hunk 2]
```diff
@@ -58,10 +61,13 @@ const PrintHeaderRow = styled(FlexContainerRow) `
...
   align-items: center;
 `;
+const TopMarginContainer = styled.div`
+  margin-top: 2em;
+`;
 
 const payloadFromState = compose(
-  filter(Boolean),
-  pick(['name', 'gender', 'email', 'birthYear', 'phoneNumber']),
+  filter(f => typeof f === 'boolean' ? true : Boolean(f)), // need to keep falsy boolean fields
```
Comment (astroash) [inline src/cb_admin/pages/Visitor.js]: falsy -> false

**Comment by astroash** [comment]
After having a little play on the site I think that the saving behaviour is different than on other pages (Activities page saves on click)

Can we have a `window.onbeforeunload` prompt before users leave the page if they haven't saved? 

https://gist.github.com/mknabe/bfcb6db12ef52323954a28655801792d

**Comment by eliascodes** [comment]
That's not related to the issue being addressed by this PR, it should be done separately.

**Pull request merged**